homeassistant/components/stream/__init__.py

```diff
@@ -95,6 +99,114 @@
 _LOGGER = logging.getLogger(__name__)
 
 
+class StreamClientError(IntEnum):
+    """Enum for stream client errors."""
+
+    BadRequest = 400
+    Unauthorized = 401
+    Forbidden = 403
+    NotFound = 404
+    Other = 4
+
+
+class StreamOpenClientError(HomeAssistantError):
+    """Raised when client error received when trying to open a stream.
+
+    :param stream_client_error: The type of client error
+    """
+
+    def __init__(
+        self, *args: Any, stream_client_error: StreamClientError, **kwargs: Any
+    ) -> None:
+        self.stream_client_error = stream_client_error
+        super().__init__(*args, **kwargs)
+
+
+async def _async_try_open_stream(
+    hass: HomeAssistant, source: str, pyav_options: dict[str, str] | None = None
+) -> InputContainer | OutputContainer:
+    """Try to open a stream.
+
+    Will raise StreamOpenClientError if an http client error is encountered.
+    """
+    return await hass.loop.run_in_executor(None, _try_open_stream, source, pyav_options)
+
+
+def _try_open_stream(
+    source: str, pyav_options: dict[str, str] | None = None
+) -> InputContainer | OutputContainer:
+    """Try to open a stream.
+
+    Will raise StreamOpenClientError if an http client error is encountered.
+    """
+    import av  # pylint: disable=import-outside-toplevel
+
+    if not pyav_options:
+        pyav_options = {}
+
+    default_pyav_options = {
+        "rtsp_flags": CONF_PREFER_TCP,
+        "timeout": str(SOURCE_TIMEOUT),
+    }
+
+    pyav_options = {
+        **default_pyav_options,
+        **pyav_options,
+    }
+
+    try:
+        container = av.open(source, options=pyav_options, timeout=5)
+
+    except av.HTTPBadRequestError as ex:
+        raise StreamOpenClientError(
+            stream_client_error=StreamClientError.BadRequest
+        ) from ex
+
+    except av.HTTPUnauthorizedError as ex:
+        raise StreamOpenClientError(
+            stream_client_error=StreamClientError.Unauthorized
+        ) from ex
+
+    except av.HTTPForbiddenError as ex:
+        raise StreamOpenClientError(
+            stream_client_error=StreamClientError.Forbidden
+        ) from ex
+
+    except av.HTTPNotFoundError as ex:
+        raise StreamOpenClientError(
+            stream_client_error=StreamClientError.NotFound
+        ) from ex
+
+    except av.HTTPOtherClientError as ex:
+        raise StreamOpenClientError(stream_client_error=StreamClientError.Other) from ex
+
+    else:
+        return container
+
+
+async def async_check_stream_client_error(
```

Comment:
Another option would be to let it raise on error and return `None` on success.  That might give us a bit more protection against having to change the API later.


```diff
diff --git a/homeassistant/components/stream/__init__.py b/homeassistant/components/stream/__init__.py
index 38c3895f4fb..496adc15279 100644
--- a/homeassistant/components/stream/__init__.py
+++ b/homeassistant/components/stream/__init__.py
@@ -186,25 +186,21 @@ def _try_open_stream(
 
 async def async_check_stream_client_error(
     hass: HomeAssistant, source: str, pyav_options: dict[str, str] | None = None
-) -> StreamClientError | None:
-    """Return an enum value representing a stream client error or None."""
-    return await hass.loop.run_in_executor(
+) -> None:
+    """Check if a stream can be successfully opened.
+
+    Raise StreamOpenClientError if an http client error is encountered.
+    """
+    await hass.loop.run_in_executor(
         None, _check_stream_client_error, source, pyav_options
     )
 
 
 def _check_stream_client_error(
-    source: str, pyav_options: dict[str, str] | None = None
-) -> StreamClientError | None:
-    """Return an enum value representing a stream client error or None."""
-    try:
-        container = _try_open_stream(source, pyav_options)
-    except StreamOpenClientError as ex:
-        return ex.stream_client_error
-    else:
-        container.close()
-
-    return None
+    source: str, pyav_options: dict[str, str] | None
+) -> None:
+    """Raise StreamOpenClientError if an http client error is encountered."""
+    _try_open_stream(source, pyav_options).close()
 
 
 def redact_credentials(url: str) -> str:
```

Reply:
That's done